pubpub/platform
# Relationship pubfield UI

## Issue(s) Resolved
Closes https://github.com/pubpub/platform/issues/617

## High-level Explanation of PR
* Adds another migration that alters the unique constraint originally introduced in https://github.com/pubpub/platform/pull/625. It turns out the one in the previous PR was not treating NULL values the way we wanted to, and to do so we have to specify additional properties when defining the unique constraint
  * Before, it was possible to insert many `{pubId: 1, fieldId: 1, relatedPubId: null}`s since psql would treat the `null`s as unique  
* Adds `Null` to the enum of `CoreSchemaType`s but does not render this as an option users can choose (yet)
* Changes the edit field modal to include a checkbox for `isRelation`. Also reorders things and renders disabled states

| Before | After |
|--------|------|
| ![image](https://github.com/user-attachments/assets/8a81e5c2-2070-4d18-b655-63133a672dc3)  |  ![image](https://github.com/user-attachments/assets/f27ae117-4b5e-4feb-bd1c-6127a6c375b5) |

* Adds playwright tests, and also fixes ones that were failing for me locally

## Test Plan
First, to test the unique constraint:

1. Run `psql postgresql://postgres:postgres@localhost:54322/postgres`
2. Run `INSERT INTO pub_values ("fieldId", value, "pubId") VALUES ('21aeb199-a586-4442-b572-6f65c29be32f', '{}', '2b056d13-6c68-48dd-81c4-6aaa38ddcc83');` (but replace your `fieldId` and `pubId` with ones that belong to a pub value that already exists in your db). You should trigger the unique constraint
3. Run `INSERT INTO pub_values ("fieldId", value, "pubId", "relatedPubId") VALUES ('21aeb199-a586-4442-b572-6f65c29be32f', '{}', '2b056d13-6c68-48dd-81c4-6aaa38ddcc83', '92b77652-f005-4aed-adb1-3571dc8495e8')`. This one should be successful
4. Run the previous command again. This one should fail by triggering the unique constraint.

To test the `isRelation` field:
1. Visit the `fields` page and create a new field.
2. If you leave the `This field represents a reference to another Pub` checkbox blank, the schema format dropdown should be required (clicking 'Create' should render an error)
3. If you check the `This field represents a reference to another Pub` checkbox, then you should be able to "Create" without error
4. The created field should render in the table with a Null schema

![image](https://github.com/user-attachments/assets/6aeea7cf-bb78-4417-9912-304935c69b66)

5. You should be able to create another relationship field with a specified schema and that should persist as well.

## Screenshots (if applicable)

New create field modal

![image](https://github.com/user-attachments/assets/f4003414-b798-49ec-8576-f6d24edce4d7)


## Notes
